server/market/market.go

```diff
@@ -587,6 +588,163 @@ func (m *Market) SubmitOrder(rec *orderRecord) error {
+// 3. Updates the storage with the new cancel order and cancels the existing limit order.
+// 4. Responds to the client that the order was received.
+// 5. Sends the unbooked order to the order feeds.
+// 6. Creates a match object, stores it, and notifies the client of the match.
+func (m *Market) processCancelOrderWhileSuspended(rec *orderRecord, errChan chan<- error) {
+	co, ok := rec.order.(*order.CancelOrder)
+	if !ok {
+		errChan <- ErrInvalidOrder
+		return
+	}
+
+	if cancelable, err := m.CancelableBy(co.TargetOrderID, co.AccountID); !cancelable {
+		errChan <- err
+		return
+	}
+
+	m.bookMtx.Lock()
+	delete(m.settling, co.TargetOrderID)
+	lo, ok := m.book.Remove(co.TargetOrderID)
+	m.bookMtx.Unlock()
+	if !ok {
+		errChan <- ErrTargetNotCancelable
+		return
+	}
+
+	m.unlockOrderCoins(lo)
+
+	sTime := time.Now().Truncate(time.Millisecond).UTC()
+	co.SetTime(sTime)
+
+	// Create the client response here, but don't send it until the order has been
+	// committed to the storage.
+	respMsg, err := m.orderResponse(rec)
+	if err != nil {
+		errChan <- fmt.Errorf("failed to create order response: %w", err)
+		return
+	}
+
+	dur := int64(m.EpochDuration())
+	now := encode.UnixMilli(time.Now())
+	epochIdx := now / dur
+	if err := m.storage.NewArchivedCancel(co, epochIdx, dur); err != nil {
+		errChan <- err
+		return
+	}
+	if err := m.storage.CancelOrder(lo); err != nil {
+		errChan <- err
+		return
+	}
+
+	err = m.auth.Send(rec.order.User(), respMsg)
+	if err != nil {
+		log.Errorf("Failed to send cancel order response: %v", err)
+	}
+
+	sig := &updateSignal{
+		action: unbookAction,
+		data: sigDataUnbookedOrder{
+			order:    lo,
+			epochIdx: 0,
+		},
+	}
+	m.sendToFeeds(sig)
+
+	match := order.Match{
+		Taker:    co,
+		Maker:    lo,
+		Quantity: lo.Remaining(),
+		Rate:     lo.Rate,
+		Epoch: order.EpochID{
+			Idx: uint64(epochIdx),
+			Dur: m.EpochDuration(),
+		},
+		FeeRateBase:  m.getFeeRate(m.Base(), m.baseFeeFetcher),
+		FeeRateQuote: m.getFeeRate(m.Quote(), m.quoteFeeFetcher),
+	}
+	// insertMatchErr is sent on errChan at the end of the function. We
+	// want to send the match request to the client even if this insertion
+	// fails.
+	insertMatchErr := m.storage.InsertMatch(&match)
+
+	makerMsg, takerMsg := matchNotifications(&match)
+	m.auth.Sign(makerMsg)
+	m.auth.Sign(takerMsg)
+	msgs := []msgjson.Signable{makerMsg, takerMsg}
+	req, err := msgjson.NewRequest(comms.NextID(), msgjson.MatchRoute, msgs)
+	if err != nil {
+		log.Errorf("Failed to create match request: %v", err)
+	} else {
+		err = m.auth.Request(rec.order.User(), req, func(_ comms.Link, resp *msgjson.Message) {
+			m.processMatchAcksForCancel(rec.order.User(), resp)
+		})
+		if err != nil {
+			log.Errorf("Failed to send match request: %v", err)
+		}
+	}
+
+	errChan <- insertMatchErr
+}
+
+// matchNotifications creates a pair of msgjson.Match from a match.
+func matchNotifications(match *order.Match) (makerMsg *msgjson.Match, takerMsg *msgjson.Match) {
```

Comment:
With returns or args of the same type, there's a shorthand where you don't need the type twice. This simplifies to:
```
func matchNotifications(match *order.Match) (makerMsg, takerMsg *msgjson.Match) {
```